ALVRClient/MetalFXUpscaler.swift

```diff
@@ -0,0 +1,86 @@
+//
+//  MetalFXUpscaler.swift
+//
+//  Created by @xuhao1 on 2024/3/16.
+//
+
+import Metal
+#if !targetEnvironment(simulator)
+import MetalFX
+#endif
+import CoreGraphics
+
+class MetalUpscaler {
+    private let device: MTLDevice
+    
+#if !targetEnvironment(simulator)
+    var mfxSpatialScaler: MTLFXSpatialScaler!
+#endif
+    var outputTexture: MTLTexture!
+    var scaling: Float32 = 1.0
+    var is_inited: Bool = false
+    
+    init?(device: MTLDevice, scaling: Float32 = 1.5) {
+        self.device = device
+        self.scaling = scaling
+    }
+    
+    func initUpscalerByTexture(inputTexture: MTLTexture) {
+        let inputWidth = inputTexture.width
+        let inputHeight = inputTexture.height
+        let pixelFormat = inputTexture.pixelFormat
+        self.setupSpatialScaler(inputWidth: inputWidth, inputHeight: inputHeight, scaling: scaling, pixelFormat: pixelFormat)
+        let textureDescriptor = MTLTextureDescriptor()
+        textureDescriptor.textureType = .type2D
+        textureDescriptor.pixelFormat = pixelFormat
+        textureDescriptor.width = Int(Float(inputWidth)*scaling)
+        textureDescriptor.height = Int(Float(inputHeight)*scaling)
+        textureDescriptor.storageMode = .private
+
+        // Optionally, set the usage of the texture. For rendering, you'll likely need shader read and render target capabilities
+        textureDescriptor.usage = [.shaderRead, .renderTarget]
+        self.outputTexture = device.makeTexture(descriptor: textureDescriptor)
+        self.is_inited = true
+        
+    }
+    
+    
+    func setupSpatialScaler(inputWidth:Int, inputHeight: Int, scaling: Float32, pixelFormat: MTLPixelFormat) {
+#if !targetEnvironment(simulator)
+        let desc = MTLFXSpatialScalerDescriptor()
+        desc.inputWidth = inputWidth
+        desc.inputHeight = inputHeight
+        desc.outputWidth = Int(Float(inputWidth)*scaling)
+        desc.outputHeight = Int(Float(inputHeight)*scaling)
+        desc.colorTextureFormat = pixelFormat
+        desc.outputTextureFormat = pixelFormat
+        desc.colorProcessingMode = .perceptual
```

Comment:
If the YUV conversion shader is changed to use `NonlinearToLinearRGB` and `EncodingNonlinearToLinearRGB`, change this to `.hdr` or `.linear`